libass/x86/be_blur.asm

```diff
@@ -31,85 +31,101 @@ SECTION .text
-.second_loop:
-    movzx r10, byte [r7 + r6] ; int temp1 = src[x];
-    lea r11, [r8 + r10] ; int temp2 = old_pix + temp1;
-    mov r8, r10 ; old_pix = temp1;
-    lea r10, [r9 + r11] ; temp1 = old_sum + temp2;
-    mov r9, r11 ; old_sum = temp2;
-    movzx r11, word [r4 + r6 * 2] ; temp2 = col_pix_buf[x];
-    add r11, r10 ; temp2 += temp1;
-    mov word [r4 + r6 * 2], r10w ; col_pix_buf[x] = temp1;
-    mov word [r12 + r6 * 2], r11w ; col_sum_buf[x] = temp2;
-    inc r6 ; x++
-    cmp r6, r1 ; x < w
-    jl .second_loop
-    mov r5, 2 ; int y = 2;
+    jl .prime_loop
+    add r8, r9 ; temp1 = old_sum + old_pix;
+    mov word [r4 + r6 * 2 - 2], r8w  ; col_pix_buf[x-1] = temp1;
+    mov word [r12 + r6 * 2 - 2], r8w ; col_sum_buf[x-1] = temp1;
+    mov r5, 1 ; int y = 1;
 .height_loop:
-    mov r10, r5; int tmpreg = y;
-    imul r10, r3; tmpreg *= stride;
-    lea r7, [r0 + r10] ; unsigned char *src=buf+y*stride;
-    sub r10, r3 ; tmpreg -= stride;
-    lea r13, [r0 + r10]; unsigned char *dst=buf+(y-1)*stride;
-    mov r6, 2 ; int x = 2;
+    mov r13, r7 ; dst = src
+    add r7, r3 ; src += stride;
+    mov r6, 1 ; int x = 1;
     movzx r10, byte [r7] ; temp1 = src[0];
-    movzx r11, byte [r7 + 1] ; temp2 = src[1];
-    add r10, r11; temp1 += temp2
-    movd xm0, r10d; __m128i old_pix_128 = temp2;
-    movd xm1, r11d; __m128i old_sum_128 = temp1;
+    movd xm0, r10d; __m128i old_pix_128 = temp1;
+    movd xm1, r10d; __m128i old_sum_128 = temp1;
 .width_loop:
-    movq xmm2, [r7 + r6]; __m128i new_pix = (src+x);
-    punpcklbw xmm2, xmm6 ; new_pix = _mm_unpacklo_epi8(new_pix, temp3);
-    movdqa xmm3, xmm2 ; __m128i temp = new_pix;
-    pslldq xmm3, 2 ; temp = temp << 2 * 8;
-    paddw xmm3, xmm0 ; temp = _mm_add_epi16(temp, old_pix_128);
-    paddw xmm3, xmm2 ; temp = _mm_add_epi16(temp, new_pix);
-    movdqa xmm0, xmm2 ; old_pix_128 = new_pix;
-    psrldq xmm0, 14 ; old_pix_128 = old_pix_128 >> 14 * 8;
-    movdqa xmm2, xmm3 ; new_pix = temp;
-    pslldq xmm2, 2 ; new_pix = new_pix << 2 * 8;
-    paddw xmm2, xmm1 ; new_pix = _mm_add_epi16(new_pix, old_sum_128);
-    paddw xmm2, xmm3 ; new_pix = _mm_add_epi16(new_pix, temp);
-    movdqa xmm1, xmm3 ; old_sum_128 = temp;
-    psrldq xmm1, 14 ; old_sum_128 = old_sum_128 >> 14 * 8;
-    movdqu xmm4, [r4 + r6 * 2] ; __m128i old_col_pix = *(col_pix_buf+x);
-    movdqu [r4 + r6 * 2], xmm2 ; *(col_pix_buf+x) = new_pix ;
-    movdqu xmm5, [r12 + r6 * 2] ; __m128i old_col_sum = *(col_pix_sum+x);
-    movdqa xmm3, xmm2 ; temp = new_pix;
-    paddw xmm3, xmm4 ; temp = _mm_add_epi16(temp, old_col_pix);
-    movdqu [r12 + r6 * 2], xmm3 ; *(col_sum_buf+x) = temp;
-    paddw xmm5, xmm3 ; old_col_sum = _mm_add_epi16(old_col_sum, temp);
-    psrlw xmm5, 4 ; old_col_sum = old_col_sum >> 4;
-    packuswb xmm5, xmm5 ; old_col_sum = _mm_packus_epi16(old_col_sum, old_col_sum);
-    movq qword [r13 + r6 - 1], xmm5 ; *(dst+x-1) = old_col_sum;
-    add r6, 8; x += 8;
+    %if mmsize == 32
```

Comment:
I think both branches should be merged.

Reply:
They're too different for this to be reasonable, due to all the extra cross-lane operations required for AVX2 (which are also why it's slower).

Reply:
I've done similar things in my code (for example [here](https://github.com/libass/libass/blob/master/libass/x86/blur.asm#L306)) and as I can see, at least half of instructions in both branches are the same. It's strange that it's slower, for me all that excess `vperm2i128` don't have much impact on performance. I'll look into `be_blur` assembly later then.